jobmaster/cvsJob/cvsJobMaster.go

```diff
@@ -82,87 +87,132 @@ func NewCVSJobMaster(ctx *dcontext.Context, workerID lib.WorkerID, masterID lib.
 }
 
 func (jm *JobMaster) InitImpl(ctx context.Context) (err error) {
-	if jm.syncInfo.DstHost == jm.syncInfo.SrcHost && jm.syncInfo.SrcDir == jm.syncInfo.DstDir {
-		return &errorInfo{info: "bad configure file ,make sure the source address is not the same as the destination"}
-	}
 	log.L().Info("initializing the cvs jobmaster  ", zap.Any("id :", jm.workerID))
 	jm.setStatusCode(lib.WorkerStatusInit)
-	jm.filesNum, err = jm.listSrcFiles(ctx)
+	filesNum, err := jm.listSrcFiles(ctx)
 	if err != nil {
 		return err
 	}
-	if jm.filesNum == 0 {
-		log.L().Info("no file found under the folder ", zap.Any("message", jm.syncInfo.DstDir))
+	if filesNum == 0 {
+		log.L().Panic("no file found under the folder ", zap.Any("id", jm.workerID))
 	}
-	log.L().Info(" cvs jobmaster list file success", zap.Any("id :", jm.workerID), zap.Any(" file number :", jm.filesNum))
+	log.L().Info("cvs jobmaster list file success", zap.Any("id", jm.workerID), zap.Any("file number", filesNum))
 	// todo: store the jobmaster information into the metastore
-	jm.syncFilesInfoMu.Lock()
-	defer jm.syncFilesInfoMu.Unlock()
-	for idx := 0; idx < jm.filesNum; idx++ {
-		conf := cvsTask.Config{SrcHost: jm.syncInfo.SrcHost, Idx: idx, DstHost: jm.syncInfo.DstHost, DstDir: jm.syncInfo.DstDir, StartLoc: "0"}
-		workerID, err := jm.CreateWorker(lib.CvsTask, conf, 10 /* TODO add cost */)
-		if err != nil {
-			// todo : handle the error case, should recover for failing to create worker
-			return err
+	for idx := 0; idx < filesNum; idx++ {
+		jm.jobStatus.FileInfos = append(jm.jobStatus.FileInfos, &SyncFileInfo{Idx: idx})
+		jm.syncFilesInfo[idx] = &WorkerInfo{
+			needCreate: *atomic.NewBool(true),
+			handle:     *atomic.NewUnsafePointer(unsafe.Pointer(nil)),
 		}
-		jm.syncFilesInfo[workerID] = &workerInfo{idx: idx, curLoc: 0, handle: nil}
 	}
+
+	// Then persist the checkpoint for recovery
+	// This persistence has to succeed before we set this master to normal status.
+	statusBytes, err := json.Marshal(jm.jobStatus)
+	if err != nil {
+		return err
+	}
+	err = jm.UserMetaKVClient().Store(ctx, jm.workerID, string(statusBytes))
+	if err != nil {
+		return err
+	}
+	jm.setStatusCode(lib.WorkerStatusNormal)
 	return nil
 }
 
 func (jm *JobMaster) Tick(ctx context.Context) error {
 	filesNum := 0
 	jm.counter = 0
-	if jm.getStatusCode() == lib.WorkerStatusInit {
-		err := jm.UpdateJobStatus(ctx, jm.Status())
-		if err != nil {
-			log.L().Warn("update job status failed", zap.Any("id", jm.workerID), zap.Error(err))
-		} else {
-			jm.setStatusCode(lib.WorkerStatusNormal)
-		}
+	if !jm.IsMasterReady() {
+		log.L().Info("jobmaster is not ready", zap.Any("master id", jm.workerID))
+		return nil
 	}
-	jm.syncFilesInfoMu.Lock()
-	defer jm.syncFilesInfoMu.Unlock()
-	for _, worker := range jm.syncFilesInfo {
-		if worker.handle == nil {
+
+	for idx, workerInfo := range jm.syncFilesInfo {
```

Comment:
if  a file has been finished before the master crash , the master should not recreate a worker for this file again. the jobmaster should store the file's sync status and process periodly

Reply:
stored in tick function